SM Samus sprite

RebelusQuo opened the pull request:
Two Todos left in the code, as well as just providing the opportunity to read the code to inform further discussion.

RebelusQuo committed:
Encapsulate the bit arrangement of tilemap entries

A class `Tilemap` is used to compile the data for a tilemap entry into the binary 5 byte data. As just one benefit, this allows the code to do direct arithmetic on the x/y values instead of handling value and the negative flag in parallel.

The "pal" key in the json data has changed to a "r" key which holds an array on the form [palette index, priority] which correspond to the bit fields in the tilemap dword, where the palette index is a 3 bit wide field, and priority is a 2 bit wide field.

RebelusQuo committed:
Provide the SM sprite asm implementation

- bugfixes.asm, improvements.asm
    - Provides small basic changes, which in parts will be overwritten by the later changes.
- gfx.asm
    - All gfx and the crystal flash palette (since that animation has been revised).
- palettes.asm
    - Palette color corrections that follow color formulas established in the SpriteSomething application
- dmadata.asm
    - Includes the binary pregenerated dma tables.
- dma.asm, gunport.asm, spinjump.asm
    - Assembler code additions.

RebelusQuo committed:
Provide an SM sprite build script for generating binary data

The build script now recursively finds and run all `build.py` files.
Note that the batch file need a wildcard in its pattern to avoid a string concat operation that tries to run a build.py through python in *every* directory.

These files are generated by the script:
- `_main_dma_prim_table.bin`
- `_main_dma_sec_[upper,lower]_table.bin`
- `_main_dma_afp_prim_table.bin`
- `_main_dma_afp_sec_table_[DB48,B800].bin`
- `_main_dma_tilemap_prim_[upper,lower]_table_at_[index].bin`
- `_main_dma_tilemap_sec_table_[8091,83C1,83E7,8A0D].bin`
- `_main_dma_tilemaps.bin`
- `_death_dma_prim_table.bin`
- `_death_dma_sec_table.bin`

Json data key legend:
- . entry name
- .. parent entry name
- o ordinal sort order
- d pixel dimensions around center
- d+ extra dimensions for irregular gfx
- pal hex entry for fifth tilemap byte
- u sequence of hex animation / pose number pairs
- lower gfx that is forced to the lower tables

RebelusQuo committed:
Update asar to version 1.71

Also add a bash shell script for scenarios such as git bash in git4win.

tewtal commented:
Looks great!
One thing that might be a good change though would be to include the correct execution address in the segments that hook/run code. For example:

`org $D08123`

before some code should be turned into

`org $D08123`
`base $908123`

This will let ASAR know that the code actually executes from bank 90 (due to mirroring in the HiROM mapping) and make it calculate jumps and addressing better.

RebelusQuo committed:
Drop unnecessary high bits in first dw of tilemap entries

Deerforce seemed to have used constants or macros that set bit 10 and 15 as well when setting bit 16 of the X dword of tilemap entries. These two bits should be unused by the hardware so we're skipping them in the generated data.

RebelusQuo committed:
Keep an empty /build folder

RebelusQuo commented:
I added `base` addresses, cleaned up the last two todos, and added a .gitignore for maintaining an empty build folder. I started up a compiled rom and Samus had all her bits, with the right colors.
